controllers/machine_controller.go

```diff
@@ -647,38 +647,70 @@ func (r *MachineReconciler) nodeToMachine(o client.Object) []reconcile.Request {
 	}
 
 	// Match by nodeName and status.nodeRef.name.
-	filters := []client.ListOption{
+	allFilters := []client.ListOption{
 		client.MatchingFields{clusterv1.MachineNodeNameIndex: node.Name},
 	}
+	annotationfilters := []client.ListOption{}
 
 	// Match by clusterName when the node has the annotation.
 	if clusterName, ok := node.GetAnnotations()[clusterv1.ClusterNameAnnotation]; ok {
-		filters = append(filters,
-			client.MatchingLabels{
-				clusterv1.ClusterLabelName: clusterName,
-			},
-		)
+		clusterNameFilter := client.MatchingLabels{
+			clusterv1.ClusterLabelName: clusterName,
+		}
+		annotationfilters = append(annotationfilters, clusterNameFilter)
+		allFilters = append(allFilters, clusterNameFilter)
 	}
 
 	// Match by namespace when the node has the annotation.
 	if namespace, ok := node.GetAnnotations()[clusterv1.ClusterNamespaceAnnotation]; ok {
-		filters = append(filters, client.InNamespace(namespace))
+		namespaceFilter := client.InNamespace(namespace)
+		annotationfilters = append(annotationfilters, namespaceFilter)
+		allFilters = append(allFilters, namespaceFilter)
 	}
 
 	machineList := &clusterv1.MachineList{}
 	if err := r.Client.List(
 		context.TODO(),
 		machineList,
-		filters...); err != nil {
+		allFilters...); err != nil {
 		return nil
 	}
 
 	// There should be exactly 1 Machine for the node.
-	if len(machineList.Items) != 1 {
+	if len(machineList.Items) == 1 {
+		return []reconcile.Request{{NamespacedName: util.ObjectKey(&machineList.Items[0])}}
+	}
+
+	machineList = &clusterv1.MachineList{}
+	if err := r.Client.List(
+		context.TODO(),
+		machineList,
+		annotationfilters...); err != nil {
+		return nil
+	}
+
+	// We need to match by ProviderID for new Nodes. Otherwise events emitted before
+	// a nodeRef is even set are lost.
+	// TODO (alberto): index by providerID once this is supported
+	// https://github.com/kubernetes-sigs/cluster-api/pull/4521/files#diff-a0e60d618a1d4037cd2adbb2da28ac0670f4b73b851ceff980326b330bf76e2b.
+	nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
+	if err != nil {
 		return nil
```

Comment:
I wonder if there are cases where it would be useful for debugging to log something here in case `err != nil && err != ErrEmptyProviderID`.